File: Java-code/Scaler-advanced/live-session/Assignments/61 - Queues - 2/minStack.java
import java.util.Stack;

public class minStack {
    Stack<Integer> stk = new Stack<>();
    Stack<Integer> minStk = new Stack<>();
    int minV = Integer.MAX_VALUE;
    public void push(int x) {
        stk.add(x);
        minV = Integer.min(minV, x);
        minStk.add(minV);
    }

    public void pop() {
        if(!stk.isEmpty()) {
            minStk.pop();
            stk.pop();
        }
        // Make sure to update the minV after popping
        if(!stk.isEmpty()) {
            minV = minStk.peek();
        } else {
            minV = Integer.MAX_VALUE;
        }
    }

    public int top() {
        if(stk.isEmpty()) {
            return -1;
        }
        return stk.peek();
    }

    public int getMin() {
        if(stk.isEmpty()) {
            return -1;
        }
        return minStk.peek();   
    }
}

File: Java-code/Revision2025/Advanced/59 - Stacks - 1/minStack.js
function solve() {
    // Initalize your variables here
    this.minStack = []
    this.stack = []
};
solve.prototype.push = function (e) {
    this.stack.push(e)
    if(this.minStack.length === 0) {
        this.minStack.push(e)
    } else {
        if (this.minStack[this.minStack.length - 1] < e) {
            this.minStack.push(this.minStack[this.minStack.length - 1])
        } else {
            this.minStack.push(e)
        }
    }
};
solve.prototype.pop = function () {
    this.stack.pop()
    this.minStack.pop()
};
solve.prototype.top = function () {
    return this.stack.length > 0 ? this.stack[this.stack.length - 1] : -1
};
solve.prototype.getMin = function () {
    return this.minStack.length > 0 ? this.minStack[this.minStack.length - 1] : -1
};

